internal/handshake/cipher_suite.go

```diff
@@ -71,15 +80,41 @@ func aeadChaCha20Poly1305(key, nonceMask []byte) *xorNonceAEAD {
 // xorNonceAEAD wraps an AEAD by XORing in a fixed pattern to the nonce
 // before each call.
 type xorNonceAEAD struct {
-	nonceMask [aeadNonceLength]byte
-	aead      cipher.AEAD
+	nonceMask        [aeadNonceLength]byte
+	aead             cipher.AEAD
+	hasSeenNonceZero bool
 }
 
 func (f *xorNonceAEAD) NonceSize() int        { return 8 } // 64-bit sequence number
 func (f *xorNonceAEAD) Overhead() int         { return f.aead.Overhead() }
 func (f *xorNonceAEAD) explicitNonceLen() int { return 0 }
 
+func allZeros(nonce []byte) bool {
+	for _, e := range nonce {
+		if e != 0 {
+			return false
+		}
+	}
+	return true
+}
+
 func (f *xorNonceAEAD) Seal(out, nonce, plaintext, additionalData []byte) []byte {
+	if useBoring && !f.hasSeenNonceZero {
+		f.hasSeenNonceZero = true
+		if !allZeros(nonce) {
+			f.sealZeroNonce(nonce)
+		}
+	}
+
+	return f.seal(nonce, out, plaintext, additionalData)
+}
+
+func (f *xorNonceAEAD) sealZeroNonce(nonce []byte) {
+	zeroNonce := make([]byte, len(nonce))
```

Comment:
Using `aeadNonceLength` instead of `len(nonce)` will probably save an allocation.

Reply:
Why?